gh-127001: Fix PATHEXT issues in shutil.which() on Windows

serhiy-storchaka opened the pull request:
* Name without a PATHEXT extension is only searched if the mode does not
  include X_OK.
* Support multi-component PATHEXT extensions (e.g. ".foo.bar").
* Support files without extensions in PATHEXT contains dot-only extension
  (".", "..", etc).
* Support PATHEXT extensions that end with a dot (e.g. ".foo.").


<!-- gh-issue-number: gh-127001 -->
* Issue: gh-127001
<!-- /gh-issue-number -->

serhiy-storchaka committed:
gh-127001: Fix conflict with extensionless files in shutil.which() on Windows

Name without a PATHEXT extension is only searched if the mode does not
include X_OK.

zooba commented:
Ah, nice. I approve of this logic. I have a vague feeling it was discussed at some point, but I think it makes sense to require an extension from `PATHEXT` if `X` is required.

serhiy-storchaka commented:
I'll rewrite tests tomorrow.

serhiy-storchaka committed:
Rewrite tests and support . in PATHEXT.

serhiy-storchaka committed:
Update comment.

serhiy-storchaka committed:
Add a NEWS entry.

serhiy-storchaka commented:
Some of existing tests did not make sense because they are grounded on wrong model. I replaced them. Also I refactored other tests to make them more uniform and reuse the same files and directories.

Fixed also cases when PATHEXT contains a multicomponent extension (e.g. `.foo.bar`) or dots (`.`, `..`, etc).

serhiy-storchaka committed:
Test full match with multidot extension

serhiy-storchaka committed:
Polishing.

serhiy-storchaka committed:
Remove an outdated bugfix versionchanged.

zooba commented on Lib/shutil.py:
```diff
@@ -1550,21 +1550,21 @@ def which(cmd, mode=os.F_OK | os.X_OK, path=None):
         pathext_source = os.getenv("PATHEXT") or _WIN_DEFAULT_PATHEXT
-        pathext = [ext for ext in pathext_source.split(os.pathsep) if ext]
+        pathext = pathext_source.split(os.pathsep)
+        pathext = [ext.rstrip('.') and ext for ext in pathext if ext]
```
I really dislike using `and` for value selection. I trust that you've got it right here, but I can't read it easily. Can we maybe comment?

```
# Replace ext of '.' with empty string, but keep trailing dots on '.ext.'
```

serhiy-storchaka replied:
Actually, I think that it is better to remove dots in all cases. The behavior of cmd.exe if `PATHEXT` contains `.BAT.` and there is `a.bat` in the path:

* `a` -- hangs
* `a.bat` -- found
* `a.` -- not found
* `a.bat.` -- not found

`which()` will now return the same result, except for the first case. I think this is the best option for now.

zooba replied:
Sounds good to me

zooba reviewed (commented):
One comment I'd like to see added, but otherwise looks good

serhiy-storchaka committed:
Support PATHEXT extension that ends with a dot.
